python/lzreposync/src/lzreposync/__init__.py

```diff
@@ -169,74 +293,48 @@ def main():
-                        channel_label=repo.channel_label,
-                    )
-                    try:
-                        dep_repo.verify()
-                    except GeneralRepoException as e:
-                        logging.error("__init__.py: Couldn't verify signature ! %s", e)
-                        exit(0)
-
-                    logging.debug("Importing package for repo %s", repo.repo_label)
-                    failed = import_repository_packages_in_batch(
-                        dep_repo,
-                        args.batch_size,
-                        channel,
-                        compatible_arches=compatible_arches,
-                    )
-                    logging.debug(
-                        "Completed import for repo %s with %d failed packages",
-                        repo.repo_label,
-                        failed,
-                    )
-                else:
-                    # TODO: handle repositories other than yum and deb
-                    logging.debug("Not supported repo type: %s", repo.repo_type)
-                    continue
+                all_channels = get_all_channels()
+                if not all_channels:
+                    print("No channels in the database! Leaving..")
+                    return
 
-        else:
-            logging.error("Either --url or --channel must be specified")
+                for channel in all_channels:
+                    ch_label = channel["channel_label"]
+                    ch_status = channel["status"]
+                    if ch_status == "pending" or ch_status == "failed":
+                        print(f"INFO: Start synchronizing channel: {ch_label}")
+                        # Update the channel status to 'in_progress'
+                        update_channel_status(ch_label, "in_progress")
+                        ret = _sync_channel(
+                            ch_label, args.cache, args.batch_size, args.no_errata
+                        )
+                        if isinstance(ret, int):
+                            channel["status"] = "failed"
+                            update_channel_status(ch_label, "failed")
+                            print(f"Failed synchronizing channel {ch_label}")
+                        elif isinstance(ret, list):
+                            if len(ret) == 0:
+                                # No failed repositories
+                                update_channel_status(ch_label, "done")
+                                print(f"Successfully synchronized channel {ch_label}")
+                            else:
+                                update_channel_status(ch_label, "failed")
+                                print(
+                                    f"Failed fully synchronizing channel {ch_label}. Some repos might have been synced."
+                                )
+                                _display_failed_repos(ret)
+
+            except KeyboardInterrupt:
```

Comment:
Honestly I'm still not completely sure about it. 

And I think we should also be careful about the final channel's `status`, after any interruption in general, and not leaving the channel in a blocking status like `in_progress`, because in that case it will be ignored by future syncs.

It still needs some refinements..

Reply:
may be in those cases the status should be set to `failed`?

Reply:
That's true...I was just thinking about a rare scenario in which all the repositories of a given channel get completely synced, but before updating the database status to `done`  the interruption happens. Although the probability of this happening is extremely low, so I think setting the status to `failed` should be good enough.

Now in terms of implementation, I would like to get your opinion about these two possible ways:

#### First one 
Add another `try...catch` block inside the for loop, to be able to get the current channel label being synced, and update its status:
```python
...
for channel in all_channels
    ch_label = channel["channel_label"]
    try:
        sync_channel_workflow(ch_label)
    except KeyboardInterrupt:
        update_status(ch_label, 'failed')
```
#### Second one
Initialize an empty variable above the first `try` block and set it to `None`: `curr_chnl_label=None`. Then it gets updated inside the `for` loop with the current channel label being synced:
```python
curr_chnl_label = None
try:
    ...
    for channel in all_channels:
        curr_chnl_label = channel["channel_label"]
        ....
except KeyboardInterrupt:
        update_status(curr_chnl_label, 'failed')
```
However in the latter case, we should add `None` check.

Both seem to have the same complexity, but I'm not sure what will be the best practice.

Reply:
If we are not sure about the actual status, "failed" looks appropriate to me. Better to run it twice than have bad data.

For the implementation, I don't think it makes a difference. Pick either of them :smile: